balena/utils.py

```diff
@@ -46,3 +56,182 @@ def compare(a, b):
         return -1
 
     return 0
+
+
+def merge(dict1: Dict[str, Any], dict2: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Merge two dictionaries, with conflicts resolved in favor of dict2.
+    """
+    result = deepcopy(dict1)
+    for key, value in dict2.items():
+        if (
+            key in result
+            and isinstance(result[key], dict)
+            and isinstance(value, dict)
+        ):
+            result[key] = merge(result[key], value)
+        else:
+            result[key] = value
+    return result
+
+
+def get_current_service_details_pine_expand(
+    expand_release: bool,
+) -> Dict[str, Any]:
+    return {
+        "image_install": {
+            "$select": ["id", "download_progress", "status", "install_date"],
+            "$filter": {
+                "status": {
+                    "$ne": "deleted",
+                },
+            },
+            "$expand": {
+                "image": {
+                    "$select": ["id"],
+                    "$expand": {
+                        "is_a_build_of__service": {
+                            "$select": ["id", "service_name"],
+                        },
+                    },
+                },
+                **(
+                    {
+                        "is_provided_by__release": {
+                            "$select": ["id", "commit", "raw_version"],
+                        },
+                    }
+                    if expand_release
+                    else {}
+                ),
+            },
+        },
+    }
+
+
+def get_single_install_summary(raw_data: Any) -> Any:
+    # TODO: Please compare me to node-sdk version
+    """
+    Builds summary data for an image install or gateway download
+
+    """
+
+    image = raw_data["image"][0]
+    service = image["is_a_build_of__service"][0]
+    release = None
+
+    if "is_provided_by__release" in raw_data:
+        release = raw_data["is_provided_by__release"][0]
+
+    install = {
+        "service_name": service["service_name"],
+        "image_id": image["id"],
+        "service_id": service["id"],
+    }
+
+    if release:
+        install["commit"] = release["commit"]
+
+    raw_data.pop("is_provided_by__release", None)
+    raw_data.pop("image", None)
+    install.update(raw_data)
+
+    return install
+
+
+def generate_current_service_details(raw_device: Any) -> Any:
+    # TODO: Please compare me to node-sdk version
+    groupedServices = defaultdict(list)
```

Comment:
Snake case is preferred it python. What do you think about
```suggestion
    grouped_services = defaultdict(list)
```

Reply:
I think that somehow flake8 is not screaming at me for some reason, I will look into it, but for sure we should be just using just snake_case